---
description: 'Disallow aliasing `this`.'
---

> ðŸ›‘ This file is source code, not the primary documentation location! ðŸ›‘
>
> See **https://typescript-eslint.io/rules/no-this-alias** for documentation.

Assigning a variable to `this` instead of properly using arrow lambdas may be a symptom of pre-ES6 practices
or not managing scope well.

## Examples

<!--tabs-->

### âŒ Incorrect

```js
const self = this;

setTimeout(function () {
  self.doWork();
});
```

### âœ… Correct

```js
setTimeout(() => {
  this.doWork();
});
```

## Options

## When Not To Use It

If you need to assign `this` to variables, you shouldnâ€™t use this rule.
